### PR TITLE
Adds nophase linter

### PR DESCRIPTION
**Pull request by theobarberbany**
This PR: 

- Clarifies the explanation around linter configurations and testdata subdirectories, as I found it a bit confusing when first reading. 
- Adds a nophase linter.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -164,6 +164,10 @@ It also supports the `// +kubebuilder:validation:Optional` and `// +kubebuilder:
 
 If you prefer to use the Kubebuilder markers instead, you can change the preference in the configuration.
 
+## Nophase
```
Comment (JoelSpeed) [inline README.md:167]: You've inserted this in the middle of a different linters block. Should move above the `## OptionalOrRequired` heading

File: docs/new-linter.md [hunk 2]
```diff
@@ -132,7 +132,9 @@ Basic tests can be implemented with the Go analysis test framework.
 Create a `testdata` directory in the linter package and create a structure underneath.
 Individual test files must be placed under `src` and then a subdirectory for each package.
 
-Use one package per configuration input for the linter.
+If your linter has different configurations, eg options to pass to the linter
```
Comment (JoelSpeed) [inline docs/new-linter.md:135]:
```suggestion
If your linter has different configurations, e.g. options to pass to the linter
```

File: pkg/analysis/nophase/testdata/src/a/a.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+type NoPhaseTestStruct struct {
+	// +optional
+	Phase *string `json:"phase,omitempty"` // want "field Phase: phase fields are deprecated and conditions should be preferred, avoid phase like enum fields"
+
+}
+
+type NoSubPhaseTestStruct struct {
+	// +optional
+	FooPhase *string `json:"phase,omitempty"` // want "field FooPhase: phase fields are deprecated and conditions should be preferred, avoid phase like enum fields"
```
Comment (JoelSpeed) [inline pkg/analysis/nophase/testdata/src/a/a.go:11]:
```suggestion
	FooPhase *string `json:"fooPhase,omitempty"` // want "field FooPhase: phase fields are deprecated and conditions should be preferred, avoid phase like enum fields"
```

File: pkg/analysis/nophase/testdata/src/a/a.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package a
+
+type NoPhaseTestStruct struct {
+	// +optional
+	Phase *string `json:"phase,omitempty"` // want "field Phase: phase fields are deprecated and conditions should be preferred, avoid phase like enum fields"
```
Comment (JoelSpeed) [inline pkg/analysis/nophase/testdata/src/a/a.go:5]: What about a field that didn't include phase in the Go name, but did include it in the serialised name?

File: pkg/analysis/nophase/analyzer.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		if !ok {
+			return
+		}
+
+		if sTyp.Fields == nil {
+			return
+		}
+
+		for _, field := range sTyp.Fields.List {
+			fieldName := field.Names[0].Name
```
Comment (JoelSpeed) [inline pkg/analysis/nophase/analyzer.go:53]: Perhaps should check if `field.Names` has `len > 0`, and that `field.Names[0]` is not nil, prior to getting the field name

```suggestion
			if len(field.Names) == 0 || field.Names[0] == nil {
				continue
			}

			fieldName := field.Names[0].Name
```

File: pkg/analysis/nophase/testdata/src/a/a.go (hunk 4 above)
Comment (theobarberbany) [reply pkg/analysis/nophase/testdata/src/a/a.go:5]: Hmm. I hadn't thought about that - we'd need to parse the jsontags then?

File: README.md [hunk 6]
```diff
@@ -164,6 +164,7 @@ It also supports the `// +kubebuilder:validation:Optional` and `// +kubebuilder:
 
 If you prefer to use the Kubebuilder markers instead, you can change the preference in the configuration.
 
+
```
Comment (JoelSpeed) [inline README.md:167]: What for?

File: README.md [hunk 7]
```diff
@@ -201,6 +202,11 @@ It will suggest to remove the pointer from the field, and update the `json` tag
 If you prefer not to suggest fixes for pointers in required fields, you can change the `pointerPolicy` to `Warn`.
 The linter will then only suggest to remove the `omitempty` value from the `json` tag.
 
+## Nophase
```
Comment (JoelSpeed) [inline README.md:205]: Can be in alphabetical order?

File: pkg/analysis/registry_test.go [hunk 8]
```diff
@@ -20,6 +20,7 @@ var _ = Describe("Registry", func() {
 				"jsontags",
 				"optionalorrequired",
 				"requiredfields",
+				"nophase",
```
Comment (JoelSpeed) [inline pkg/analysis/registry_test.go:23]: Should make alphabetical

**Review by JoelSpeed — DISMISSED** [review]

File: pkg/analysis/nophase/testdata/src/a/a.go (hunk 4 above)
Comment (theobarberbany) [reply pkg/analysis/nophase/testdata/src/a/a.go:5]: we now parse jsontags

**Pull request merged**